### ISSUE TITLE
Tests failing on Windows PHP 5.6.13

### ISSUE THREAD

**Issue by garethellis36** [thephpleague/commonmark#189]
Hello,

7 of the tests are failing under PHP 5.6.13 for Windows. I cloned the repository, ran `composer install` and then `phpunit`. This may be related, but in addition, if I run `php bin\commonmark`, nothing seems to happen (I have to CTRL-C out of PHP).

Results:

```shell
C:\inetpub\commonmark>phpunit
PHPUnit 4.7.7 by Sebastian Bergmann and contributors.

Runtime:        PHP 5.6.13 with Xdebug 2.3.2
Configuration:  C:\inetpub\commonmark\phpunit.xml.dist

FFFFFFF.......................................................  62 / 956 (  6%)
.............................................................. 124 / 956 ( 12%)
.............................................................. 186 / 956 ( 19%)
.............................................................. 248 / 956 ( 25%)
.............................................................. 310 / 956 ( 32%)
.............................................................. 372 / 956 ( 38%)
.............................................................. 434 / 956 ( 45%)
.............................................................. 496 / 956 ( 51%)
.............................................................. 558 / 956 ( 58%)
.............................................................. 620 / 956 ( 64%)
.............................................................. 682 / 956 ( 71%)
.............................................................. 744 / 956 ( 77%)
.............................................................. 806 / 956 ( 84%)
.............................................................. 868 / 956 ( 90%)
.............................................................. 930 / 956 ( 97%)
..........................

Time: 3.43 minutes, Memory: 47.00Mb

There were 7 failures:

1) League\CommonMark\Tests\Functional\BinTest::testNoArgsOrStdin
Failed asserting that 'Failed without error message: C:\inetpub\commonmark\bin\commonmark' contains "Usage:".

C:\inetpub\commonmark\tests\functional\BinTest.php:19

2) League\CommonMark\Tests\Functional\BinTest::testHelpShortFlag
Failed asserting that 1 matches expected 0.

C:\inetpub\commonmark\tests\functional\BinTest.php:31

3) League\CommonMark\Tests\Functional\BinTest::testHelpOption
Failed asserting that 1 matches expected 0.

C:\inetpub\commonmark\tests\functional\BinTest.php:44

4) League\CommonMark\Tests\Functional\BinTest::testUnknownOption
Failed asserting that 'Failed without error message: C:\inetpub\commonmark\bin\commonmark --foo' contains "Unknown option".

C:\inetpub\commonmark\tests\functional\BinTest.php:58

5) League\CommonMark\Tests\Functional\BinTest::testFileArgument
Failed asserting that 1 matches expected 0.

C:\inetpub\commonmark\tests\functional\BinTest.php:70

6) League\CommonMark\Tests\Functional\BinTest::testStdin
Failed asserting that 1 matches expected 0.

C:\inetpub\commonmark\tests\functional\BinTest.php:83

7) League\CommonMark\Tests\Functional\LocalDataTest::testExample with data set #0 ('# H1\n\n## H2\n\n### H3\n\n#### H4\n...#####\n', '<h1>H1</h1>\n<h2>H2</h2>\n<h3>H...</h6>\n', 'atx_header')
Unexpected result for "atx_header" test
=== markdown ===============
# H1

## H2

### H3

#### H4

##### H5

###### H6

# H1 #

## H2 ##

### H3 ###

#### H4 ####

##### H5 #####

###### H6 ######

=== expected ===============
<h1>H1</h1>
<h2>H2</h2>
<h3>H3</h3>
<h4>H4</h4>
<h5>H5</h5>
<h6>H6</h6>
<h1>H1</h1>
<h2>H2</h2>
<h3>H3</h3>
<h4>H4</h4>
<h5>H5</h5>
<h6>H6</h6>

=== got ====================
<h1>H1</h1>
<h2>H2</h2>
<h3>H3</h3>
<h4>H4</h4>
<h5>H5</h5>
<h6>H6</h6>
<h1>H1</h1>
<h2>H2</h2>
<h3>H3</h3>
<h4>H4</h4>
<h5>H5</h5>
<h6>H6</h6>

Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
 '<h1>H1</h1>
 <h2>H2</h2>
 <h3>H3</h3>
 <h4>H4</h4>
 <h5>H5</h5>
 <h6>H6</h6>
 <h1>H1</h1>
 <h2>H2</h2>
 <h3>H3</h3>
 <h4>H4</h4>
 <h5>H5</h5>
 <h6>H6</h6>
 '

C:\inetpub\commonmark\tests\functional\LocalDataTest.php:52

FAILURES!
Tests: 956, Assertions: 1312, Failures: 7.

Generating code coverage report in Clover XML format ... done

Generating code coverage report in HTML format ... done
```

**Comment by garethellis36** [comment]
By running `php bin\commonmark` and judiciously adding 'die()', I worked out that the script is getting 'stuck' at line 36 `$markdown = stream_get_contents($stdin);`.

**Comment by garethellis36** [comment]
The problem is that `stream_set_blocking($stdin, false);` returns false on Windows and doesn't really work. There's a whole bunch of PHP bugs that are really old - see [man page](http://php.net/manual/en/function.stream-set-blocking.php).

In terms of the CLI for this package, it's not a big deal - maybe a README note that if you just run "commonmark" without any input on Windows it will hang and you need to quit with CTRL-C? I tested with input and it works as expected.

The first test is failing presumably because of this issue on Windows, I haven't yet figured out why the others are failing. I'll take another look tomorrow.

**Comment by colinodell** [comment]
Once challenge I faced on Linux (before adding `stream_set_blocking`) was that `STDIN` blocks indefinitely until it receives input; you therefore cannot possibly show the help if nothing is piped through.

Ideally I'd like to support `STDIN` on Windows. However, if this isn't possible, I'm totally okay with disabling this feature for Windows.

**Comment by colinodell** [comment]
Perhaps we should just skip the whole `STDIN` reading part if `stream_get_contents()` returns `false`?  That would properly workaround this issue on Windows.

**Comment by GrahamCampbell** [comment]
Are the others failing because of the line endings?

**Comment by garethellis36** [comment]
@colinodell problem is `stream_get_contents()` doesn't return false on Windows, it just hangs PHP at that point. `stream_set_blocking($stdin, false)` returns false indicating that blocking mode is still in action. STDIN does work, it's only when you run it with no input at the CLI:

```shell
C:\inetpub\commonmark>echo '# Hello World!' | php bin\commonmark
<p>'# Hello World!'</p>```

vs.

```shell
C:\inetpub\commonmark>php bin\commonmark
//hangs, have to end with CTRL-C
```

**Comment by garethellis36** [comment]
Other tests are failing because the structure of `BinTest.php` sends the command "commonmark" to the `Command` object. On Windows, if you haven't configured Windows to be able to run PHP files as scripts (I haven't done this) then this will not work. So the test are failing because it's not actually running the script through PHP - Windows is silently failing with a `'bin\commonmark' is not recognized as an internal or external command, operable program or batch file`-style error.

The fix for this is simple - prefixing the call to `realpath()` in `BinTestgetPathToCommonmark()` with 'php':
```php
    protected function getPathToCommonmark()
    {
        return "php " . realpath(__DIR__ . '/../../bin/commonmark');
    }
```
Now running `phpunit BinTest` sees the same PHP 'hanging' problem described above.

**Comment by colinodell** [comment]
@garethellis36 Sorry for the delay in getting to this.  I've implemented a possible solution via #195 - would you mind testing this when you get a chance?
